crates/polars-io/src/ipc/object_store_reader.rs

```diff
@@ -0,0 +1,87 @@
+use bytes::Bytes;
+use object_store::path::Path as ObjectPath;
+use object_store::ObjectStore;
+use polars_core::frame::DataFrame;
+use polars_error::{to_compute_err, PolarsResult};
+
+use crate::pl_async::{get_runtime, with_concurrency_budget};
+use crate::prelude::IpcReader;
+use crate::SerReader;
+
+pub trait ObjectStoreIpc {
+    fn read_ipc(&self, path: &ObjectPath) -> PolarsResult<DataFrame>;
+}
+
+impl<T> ObjectStoreIpc for T
+where
+    T: ObjectStore,
+{
+    fn read_ipc(&self, path: &ObjectPath) -> PolarsResult<DataFrame> {
+        // TODO: This will block the current thread until the data has been
+        // loaded. No other (compute) work can be done on this thread.
+        //
+        // What is our work scheduling strategy? From what I have heard, we do
+        // not intend to make the entire library `async`. However, what do we do
+        // instead?
+        get_runtime().block_on(read_ipc_inner(self, path))
+    }
+}
+
+async fn read_ipc_inner<T: ObjectStore>(store: &T, path: &ObjectPath) -> PolarsResult<DataFrame> {
+    // TODO: Load only what is needed, rather than everything.
+    let file_bytes = with_concurrency_budget(1, || read_bytes(store, path)).await?;
+
+    let reader = IpcReader::new(std::io::Cursor::new(file_bytes));
+
+    let data = reader.finish()?;
+
+    PolarsResult::Ok(data)
+}
+
+async fn read_bytes<T: ObjectStore>(store: &T, path: &ObjectPath) -> PolarsResult<Bytes> {
```

Comment:
This must eventually launch a thread that prefetches `ArrowChunk` and stores them on a queue.

The `AsyncIpcReader` must have a `batched` method that turns this reader into a `BatchedReader`.

The `BatchedReader` will have a `next_batches` that will fetch batches from that queues receiving end.

You can mostly copycat all fields/methods from parquet. E.g. we need to be able to apply a `limit`, project only a certain amount of columns, apply a filter etc.